### ISSUE TITLE
[Feature Request] Folder Preview 

### ISSUE THREAD

**Issue by Gigi33** [arpitkh96/AmazeFileManager#315]
Would be nice to have a quick way of seeing what's inside a folder before opening it. Usually the first few thumbnails of the folder content are shown in a combined fashion.

I'd love to see the the possibility of configuring this feature as well.  Things like number of thumbs in the preview, style of preview (stack/grid/fan/line), etc. 

Would this be merged upstream if I would implement it and create a pull request?

**Comment by vishal0071** [comment]
I'm sorry for late reply. If you're still interested, we can discuss on the same.

**Comment by Gigi33** [comment]
Yeah, I would be interested in the feature!

**Comment by vishal0071** [comment]
So, what's the idea? How would you like a preview to be shown? Longpress is registered for selection, icon press is registered for selection too (I'm think to remake it to view properties or something) then the three-dot options menu for options. How else can we show a preview? Single click is not an option, as I won't be increasing the no.of clicks required to access data and ruin the UX.

**Comment by Gigi33** [comment]
I was thinking along the lines of showing folder previews directly. A good example are the launch folders of the Nova Launcher, which is where I've seen the [different styles of preview (stack/grid/fan/line)](http://i.imgur.com/GQ23XP7.png).

There are [multiple examples](http://i.imgur.com/CLqQLBD.png) of how something like that could look - I'm not too sure myself of what is feasible / what looks nice. 

There is the [Windows Explorer style of previews](http://i.imgur.com/G5FobD6.jpg) as well, of course. Or some [plain old tiling](http://i.imgur.com/BLs2q1U.png).

I'm sure you can find [prettier examples](http://i.imgur.com/4bgFVXO.jpg) of this idea if you look around. I haven't done any research yet, but I'm sure there are some nice concept designs out there. Not sure if it works as well on mobile though.

**Comment by Gigi33** [comment]
On the technical side I have no idea what's feasible on mobile devices. In terms of performance one could load the thumbnail embedded in the EXIF data. I guess generating the thumbnails on-the-fly could be quite expensive. For photos the EXIF thumbnail should work quite well, not sure what to do with movies or other formats that could benefit from a visual preview. I mean you could always set a limit if you don't want to generate thumbnails for huge files.

Thoughts?

**Comment by vishal0071** [comment]
I really appreciate your reply with examples. What I've come up is, adding this as one of a feature of grid view. Inspiration from Nova/Windows style stack of icons, the grid view's big size icon is feasible for this feature. However, the listview icons are too small to be grouped. Hence may make UX really hefty and bad. What say?

**Comment by Gigi33** [comment]
Yeah, grid view would be great! I doubt as well that it would make sense in the list view. 

I'm currently quite swamped with my day job, but maybe I'll have a few minutes on the weekend so I can whip up some mockups in Photoshop to see what looks/feels right.

**Comment by vishal0071** [comment]
Alright! I appreciate your support. Meanwhile I'll think of a good design too.

**Comment by Gigi33** [comment]
Hey, I whipped up something quickly. It's very basic. Three styles:
1. A grid of circles
2. A grid of squares
3. Single-preview masked by the folder icon

![amaze-grid-preview](https://cloud.githubusercontent.com/assets/109058/19619913/babfe3d2-9871-11e6-8180-bd0a2448e223.png)

I have some more ideas that I'd like to try out visually first - I'm in a hurry unfortunately. Maybe I'll have time tomorrow. 

I don't think that the limit space we have to work with allows for more than 3 previews. Maybe on tablets or humongous devices it makes sense, I don't know.

What do you think?

**Comment by vishal0071** [comment]
Second one seems really good! I see upto 4 previews in your case, but it can be modified based on the density of the device, so that's not the concern, let me handle that part. I would like you to specifically try the Windows style folder, let's see if that looks good.

**Comment by Gigi33** [comment]
Sounds good! I'll also do some research in regards to embedded metadata in
video formats and the like.

**Comment by Gigi33** [comment]
Hey, sorry for the late reply - finally got around to whip up the Windows-style preview.

I copied the Win10 style pretty much exactly. Just changed the color of the folders to match the one in the screenshot. There are - of course - multiple folder styles. One could also display it [like this](https://i.imgur.com/bjk0PeE.jpg), for example.

![amaze-win10-style-preview](https://cloud.githubusercontent.com/assets/109058/19910077/25157c44-a08b-11e6-8f02-ca2ffb720343.png)

What do you think?

**Comment by vishal0071** [comment]
Looking so beautiful! Thanks for your time. I'm stuck with some other critical bugs, but after those, I definitely have this in priority.

**Comment by Gigi33** [comment]
I did some research in regards to video thumbnails. 

As mentioned before, for photos we could use the thumbnails embedded in the EXIF metadata (if the image file has EXIF metadata). The thumbnail would come "for free" so to speak, and we wouldn't have to generate it. For non-photo images we would have to generate it anyway, though.

For videos it is a different story it seems. There is no real standard since there are so many different video containers and video formats. I experimented with [exiftool](http://www.sno.phy.queensu.ca/~phil/exiftool/) since it can read most video containers, but I had no luck extracting thumbnails from my videos. 

`exiftool -b -ThumbnailImage VIDEO.mp4 > thumbnail.jpg` did not seem to do the trick unfortunately.

However, it is quite easy to generate video thumbnails using [ffmpeg](https://www.ffmpeg.org/).

`ffmpeg -i VIDEO.mp4 -ss 00:00:01.000 -vframes 1 thumbnail.png` works like a charm.

I'm sure there a plenty of Java wrappers for `ffmpeg`, so even though it might be costly computationally speaking we could do it this way. Not sure where to store the generated thumbnails, though. Approaches of file managers and other applications differ quite a bit, and reach from hidden metadata folders (e.g. OS X Finder) or hidden metadata files to storing the thumbnails in a database (e.g. Picasa, as far as I can tell). I guess you already have something in mind, just wanted to share my thoughts.

**Comment by Lokesh-Krishna** [comment]
As a regular user who loves the list view, could the single preview masked by the folder icon make it was well?

**Comment by EmmanuelMess** [comment]
This would have to be added to glide, some system to create an image from multiple images.
